src/dune_engine/build_system.ml

```diff
@@ -2008,27 +2008,24 @@ end = struct
           Path.Build.Map.find_exn targets path)
 
   let build_alias_impl alias =
-    expand_alias_gen alias ~eval_build_request:exec_build_request
-      ~paths_of_facts:(fun facts -> Memo.Build.return (Dep.Facts.paths facts))
-      ~paths_union_all:(fun l ->
-        List.fold_left l ~init:Path.Map.empty
-          ~f:(Path.Map.union ~f:(fun _ x _ -> Some x)))
-
-  module Digest_path_map = struct
-    type t = Digest.t Path.Map.t
-
-    let to_dyn = Path.Map.to_dyn Digest.to_dyn
-
-    let equal = Path.Map.equal ~equal:Digest.equal
-  end
+    let+ files =
+      expand_alias_gen alias ~eval_build_request:exec_build_request
+        ~paths_of_facts:(fun facts -> Memo.Build.return (Dep.Facts.paths facts))
+        ~paths_union_all:(fun l ->
```

Comment:
I'm guessing this `paths_union_all` operation may still carry similar unnecessary costs: if an alias is included in multiple aliases then you pay for every file per every alias it's included in.

That's where the tree nature of Proxy_map may be beneficial.

Anyway, the PR should clearly improve the common case.